src/snowflake/snowpark/_internal/utils.py

```diff
@@ -703,3 +703,72 @@ def strip_double_quotes_in_like_statement_in_table_name(table_name: str) -> str:
 
     # if table_name == '"a.b"', then we should call SHOW TABLES LIKE 'a.b'
     return table_name[1:-1] if table_name[0] == table_name[-1] == '"' else table_name
+
+
+def parse_table_name(table_name: str) -> List[str]:
+    """
+    This function implements the algorithm to parse a table name.
+
+    We parse the table name according to the following rule:
+        https://docs.snowflake.com/en/sql-reference/identifiers-syntax
+        Unquoted object identifiers:
+        Start with a letter (A-Z, a-z) or an underscore (“_”).
+        Contain only letters, underscores, decimal digits (0-9), and dollar signs (“$”).
+        Are stored and resolved as uppercase characters (e.g. id is stored and resolved as ID).
+        If you put double quotes around an identifier (e.g. “My identifier with blanks and punctuation.”),
+        the following rules apply: The case of the identifier is preserved when storing and resolving
+        the identifier (e.g. "id" is stored and resolved as id).
+        The identifier can contain and start with ASCII, extended ASCII, and non-ASCII characters.
+        To use the double quote character inside a quoted identifier, use two quotes. For example:
+    """
+    validate_object_name(table_name)
+    str_len = len(table_name)
+    ret = []
+
+    in_double_quotes = False
+    i = 0
+    cur_word_start_idx = 0
+
+    while i < str_len:
+        cur_char = table_name[i]
+        if cur_char == '"':
+            if in_double_quotes:
+                # we have to check whether this `"` is the ending of a double-quoted identifier
+                # or it's an escaping double quote
+                # to achieve this, we need to preload one more char
+                if i == str_len - 1:
```

Comment:
I feel we should not blend double quote check with end-of-string check. How about the following:

```
if i < str_len - 1 and table_name[i + 1] == '"':
  i += 1
else:
  in_double_quotes = False
```

This way we keep the end of string check in one place.

Reply:
nice catch, I think the logic indeed can be merged, let me update